File: backend/app/schemas/library.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.base import BaseSchema

Visibility = Literal["board", "members"]


# --- Folders ---


class LibraryFolderCreate(BaseSchema):
    name: str = Field(min_length=1, max_length=255)
    parent_id: uuid.UUID | None = None
    sort_order: int = 0


class LibraryFolderUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    # Explicit `null` moves the folder to the root, which is why every write
    # path reads this with `exclude_unset=True`: "not mentioned" and "moved to
    # the top" are different requests and must not collapse into one.
    parent_id: uuid.UUID | None = None
    sort_order: int | None = None


class LibraryFolderResponse(BaseSchema):
    id: uuid.UUID
    parent_id: uuid.UUID | None = None
    name: str
    sort_order: int
    created_at: datetime
    updated_at: datetime


# --- Documents ---


class LibraryDocumentUpdate(BaseSchema):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=5000)
    folder_id: uuid.UUID | None = None
    visibility: Visibility | None = None


class LibraryDocumentResponse(BaseSchema):
    id: uuid.UUID
    folder_id: uuid.UUID | None = None
    title: str
    description: str | None = None
    visibility: Visibility
    original_filename: str
    content_type: str
    byte_size: int
    checksum_sha256: str
    uploaded_by_user_id: uuid.UUID | None = None
    uploaded_at: datetime
    replaces_id: uuid.UUID | None = None
    superseded_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class LibraryUsageResponse(BaseSchema):
    """What the club has used and what it may use.

    The upload form needs all three before it lets someone pick a 40 MB scan
    and watch it fail at the end.
    """

    used_bytes: int
    quota_bytes: int
    max_upload_bytes: int

File: apps/web/lib/types/library.ts
/**
 * The club's filing cabinet — "Ablage" in the UI, `library` in the code.
 *
 * Not to be confused with `lib/types/document.ts`, which is the certificate
 * module: those documents are *produced* from a template, these are files
 * somebody uploaded.
 */

export type LibraryVisibility = "board" | "members"

export type LibraryFolder = {
  id: string
  parent_id: string | null
  name: string
  sort_order: number
  created_at: string
  updated_at: string
}

export type LibraryDocument = {
  id: string
  folder_id: string | null
  title: string
  description: string | null
  visibility: LibraryVisibility
  original_filename: string
  content_type: string
  byte_size: number
  checksum_sha256: string
  uploaded_by_user_id: string | null
  uploaded_at: string
  replaces_id: string | null
  superseded_at: string | null
  created_at: string
  updated_at: string
}

export type LibraryUsage = {
  used_bytes: number
  quota_bytes: number
  max_upload_bytes: number
}
